pyop2/fusion.py

```diff
@@ -0,0 +1,1264 @@
+                    else:
+                        # Neither READ nor WRITE, so access modes are some
+                        # combinations of RW, INC, MIN, MAX. For simplicity,
+                        # just make it RW.
+                        filtered_args[a.data]._access = RW
+        return filtered_args
+
+    def c_arg_bindto(self, arg):
+        """Assign c_pointer of this Arg to ``arg``."""
+        if self.ctype != arg.ctype:
+            raise RuntimeError("Cannot bind arguments having mismatching types")
+        return "%s* %s = %s" % (self.ctype, self.c_arg_name(), arg.c_arg_name())
+
+    def c_map_name(self, i, j):
+        return self._c_local_maps[i][j]
+
+    @property
+    def name(self):
+        """The generated argument name."""
+        return "arg_exec_loop%d_%d" % (self._loop_position, self._position)
+
+
+class Kernel(host.Kernel, tuple):
+
+    """A :class:`fusion.Kernel` object represents an ordered sequence of kernels.
+    The sequence can either be the result of the concatenation of the kernels
+    bodies, or a list of separate kernels (i.e., different C functions).
+    """
+
+    @classmethod
+    def _cache_key(cls, kernels, fused_ast=None, loop_chain_index=None):
+        keys = "".join([super(Kernel, cls)._cache_key(k._code or k._ast.gencode(),
+                                                      k._name, k._opts, k._include_dirs,
+                                                      k._headers, k._user_code)
+                        for k in kernels])
+        return str(loop_chain_index) + keys
+
+    def _ast_to_c(self, asts, opts):
+        """Fuse Abstract Syntax Trees of a collection of kernels and transform
+        them into a string of C code."""
+        if not isinstance(asts, (ast.FunDecl, ast.Root)):
+            asts = ast.Root(asts)
+        self._ast = asts
+        self._original_ast = dcopy(asts)
+        return super(Kernel, self)._ast_to_c(self._ast, opts)
+
+    def __init__(self, kernels, fused_ast=None, loop_chain_index=None):
+        """Initialize a :class:`fusion.Kernel` object.
+
+        :param kernels: an iterator of some :class:`Kernel` objects. The objects
+                        can be of class `fusion.Kernel` or of any superclass.
+        :param fused_ast: the Abstract Syntax Tree of the fused kernel. If not
+                          provided, kernels are simply concatenated.
+        :param loop_chain_index: index (i.e., position) of the kernel in a loop
+                                 chain. This can be used to differentiate a same
+                                 kernel appearing multiple times in a loop chain.
+        """
+        # Protect against re-initialization when retrieved from cache
+        if self._initialized:
+            return
+
+        asts = fused_ast
+        if not asts:
+            # If kernels' need be concatenated, discard duplicates
+            kernels = dict(zip([k.cache_key[1:] for k in kernels], kernels)).values()
```

Comment:
We are just discarding duplicate code. Of course the same kernel can appear multiple times in the fused parloop: this is totally supported, but this "discarding thing" is just to avoid that the same piece of C code uselessly appears multiple times